0000/README.adoc

```diff
@@ -0,0 +1,175 @@
+
+https://jcp.org/en/jsr/detail?id=222[Bean Validation] is an annotation-based framework designed for
+generic usage defining constraints validation on data in a Java model. It allows definition of model
+constraints in a fully declarative way, and define the mechanism to run validation and discover violations.
+
+Bean Validation also is extensible by developers who can define custom annotations and rules to implement
+them.
+
+==== Relying on JSON-B
+
+http://json-b.net/[Json-B] (JSR 367) is the approved specification for JSON binding. It has been highly
+inspired by https://github.com/FasterXML/jackson[Jackson] and do define the exact binding requirements
+Stapler do implement, but relying on standard annotations (vs custom `@DataBoundConstructor`.
+
+As Stapler will still need to support legacy data-binding for backward compatibility, we can't just replace
+it with a Json-B implementation. We could consider adding support for such standard annotations as
+alternatives to stapler specific ones. But on the other way a Jenkins databound component is nothing like
+a reusable piece of software, and the https://javadoc.io/doc/javax.json.bind/javax.json.bind-api/1.0[Json-B annotations]
+are mostly designed to tune Java to Json conversion, not providing huge benefits.
+
+==== Filed vs Setter injection
+
+An endless debate for annotation based framework is about using annotations on private fields, vs using them
+on accessors. Most frameworks do support both as there's no single answer to this debate.
+
+On one side, injection on private field require Java reflexion to unlock private field accessibility (or
+use Variable Handles on java 9+). It makes testing harder as there's no trivial way to mock or instantiate
+the target component.
+
+On the other side, setter injection require some boilerplate code being added to codebase, and don't prevent
+external component to create an instance without invoking such setters, resulting in misconfigured component.
+Same applies to any `@PostConstruct` initializer method.
+
+Generally speaking, a component designed for data-binding can't guarantee it exposes a safe API until there's just
+a single constructor to require and validate all attributes. From this point of view, we consider a web UI
+databound component in Jenkins is nothing but a reusable component, and as such should never be used outside this
+very specific context.
+
+Based on this:
+
+1. databinding annotations can be used both on fields and accessors
+1. elements annotated databinding annotation will be automatically marked a restricted so they are not exposed as
+"public API" whatever their Java visibility is.
+
+== Backwards Compatibility
+
+Annotation based databinding comes in addition to the legacy mechanisms supporter by stapler, so don't break
+backward compatibility.
+
+== Security
+
+N/A
+
+== Infrastructure Requirements
+
+N/A
+== Testing
+
+N/A
+
+== Prototype Implementation
+
+https://github.com/stapler/stapler/pull/140
```

Comment:
Not enough; you need to make `structs` work with it. Also there is a bit of miscellaneous code in `workflow-cps` (`Snippetizer`) that would probably require integration on top of what it picks up from `structs`.

And of course a reference implementation should include a broad range of uses of the new model, as a convincing demonstration that it can be used to solve real problems and simplify code. Preferably in existing code bases (core, important plugins) where possible to do so without breaking compatibility too badly.